quill-finagle-mysql/src/main/scala/io/getquill/sources/finagle/mysql/FinagleMysqlSource.scala

```diff
@@ -53,12 +44,12 @@ class FinagleMysqlSource[N <: NamingStrategy](config: FinagleMysqlSourceConfig[N
   def probe(sql: String) =
     Try(Await.result(client.query(sql)))
 
-  def transaction[T](f: => Future[T]) =
-    client.transaction {
-      transactional =>
-        currentClient.update(transactional)
-        f.ensure(currentClient.clear)
+  def transaction[T](f: FinagleMysqlSource[N] => Future[T]) = {
+    val source = new FinagleMysqlSource(config)
+    source.client.transaction { tx =>
+      f(source)
```

Comment:
The same approach I've suggested for `AsynSource` could be used here. This won't make the future run within the transaction.